File: old/calculateCurrent.m
function [I, pointsx, pointsy] = calculateCurrent(x, y, params)

L = params.L;
voltage = params.V;
lambda = params.lambda;
Rt = params.Rt;

p = 1000;
t2 = 1000;

sz = size(x);
t1 = sz(1);
n = sz(2);
newL = 200e-6;

pointsx = zeros(t1,n);
pointsy = zeros(t1,n);
I = zeros(t1,1);

for i = 1:t1
    
    count2 = 1;
    
    % Start with electrons at random particles
    
%     curr_indexes = randsample(1:n,p,true);
%     currx = x(i,curr_indexes);
%     curry = y(i,curr_indexes);

    currx = zeros(1,p);
    curry = zeros(1,p);
    curr_indexes = zeros(1,p);
    
    xi = x(i,:);
    yi = y(i,:);
    
    for j = 1:t2
        
        % a1 = ~ismember(currx(:),pointsx(i,:));
        % a2 = ~ismember(curry(:),pointsy(i,:));
        
        % a3 = find(a1 == 1 & a2 == 1);
        
        % if ~isempty(a3)
        %     count = length(a3);
        %     pointsx(i,count2:count2+count-1) = currx(a3);
        %     pointsy(i,count2:count2+count-1) = curry(a3);
        %     count2 = count2 + count;
        % end
            
        % Get distances to particles in +x direction
        
        dx = repmat(xi,p,1) - currx';
        dy = repmat(yi,p,1) - curry';
        
        dy(dy==repmat(yi,p,1)) = 0;
        
        d = sqrt(dx.^2 + dy.^2);
        
        positive = dx >= 0;
        
        dist_weights = positive .* 1./(Rt*exp(d/lambda));
        dist_weights(isnan(dist_weights)) = 0;
        
        to_end = (1./(Rt*exp((newL-currx')/lambda))) .* ones(p,1);
        to_end(isnan(to_end)) = 1;
        
        stay_at_beginning = currx' == 0;
        
        tot_weights = [dist_weights stay_at_beginning to_end];
        
        [curr_indexes,currx,curry]=getNextIndex(i,tot_weights,x,y,currx,curry,curr_indexes);
        
    end
    I(i) = sum(isnan(currx));
end

%     sz = size(x);
%     t1 = sz(1);
% 
%     n = params.n;
%     newL = 200e-6;
% 
%     pointsx = zeros(t1,n);
%     pointsy = zeros(t1,n);
%     I = zeros(t1,1);
% 
%     for i = 1:t1
% 
%         count2 = 1;
% 
%         xi = x(i,:);
%         yi = y(i,:);
% 
%         % Get initial electron positions
%         good = (xi > 0) .* (xi < L);
%         curr_indexes = find(good == 1);
%         w = ones(1,length(curr_indexes));
% 
%         if length(curr_indexes) > 1
%             curr_indexes = randsample(curr_indexes, num_e, true, w);
%         end
%         if length(curr_indexes) > num_e
%             curr_indexes = curr_indexes(1:num_e);
%         end
% 
%         p = length(curr_indexes);
%         if p == 0
%             break
%         end
% 
%         % Start with electrons at all particles in device
%         directions = [1,-1];
%         for dirNum=1:2
%             % Get direction value
%             direction = directions(dirNum);
% 
%             % Get initial indices
%             curr_indexes_local = curr_indexes;
%             currx_local = xi(curr_indexes_local);
%             curry_local = yi(curr_indexes_local);
% 
%             for j = 1:steps
% 
%                 % Get distances
%                 dx = repmat(xi,p,1) - currx_local';
%                 dy = repmat(yi,p,1) - curry_local';
%                 d = sqrt(dx.^2 + dy.^2);
%                 dx = dx*direction;
%                 positive = dx >= 0;
% 
%                 % Get resistances
%                 voltage = V*dx/L;
%                 sigmoid = 1./(1+exp(-voltage));
%                 resist = Rt*exp((d)/lambda);
% 
%                 dist_weights = positive .* sigmoid./resist;
%                 dist_weights(isnan(dist_weights)) = 0;
% 
%                 d = direction*(L/2 - currx_local') + L/2;
% 
%                 % Get resistance to end
%                 voltage_end = V.*(d)./L;
%                 sigmoid_end = 1./(1+exp(-voltage_end));
%                 resist_end = Rt*exp((d)/lambda);
% 
%                 to_end = (sigmoid_end./resist_end) .* ones(p,1);
%                 to_end(isnan(to_end)) = 1;
%                 to_begin = zeros(p,1);
% 
%                 tot_weights = [dist_weights to_begin to_end];
% 
%                 [curr_indexes_local,currx_local,curry_local] = getNextIndex(tot_weights, x, y, currx_local, curry_local, curr_indexes_local);
%             end
%         end
% 
%         % Add plots
%         I(i) = sum(isnan(currx_for) .* isnan(currx_back));
%         % disp(I(i))
%         % hold onx
%         % plot(i,I(i),'b.');
%         % drawnow
%     end
% end

File: calculateCurrent.py
from defineParameters import params
from numpy import exp, ones, sum, copy, max, array, argmin, sqrt, argsort, hstack, zeros, float32 #tile, full, isnan, sqrt, min, nan, newaxis
from numpy.random import rand
from numpy import append as app
#from getNextIndex import get_next_index

def distances(x1, x2, y1, y2):
    dx = x1 - x2
    dy = y1 - y2
    d = sqrt(dx ** 2 + dy ** 2)
    return d

def boltz_fact(x_i,x_j,Volt, T, L_x):
    '''
    x_i is the current location of the electron and x_j is the location of the electron in the proposed step
    '''
    V_i = Volt * (x_i/L_x)
    V_j = Volt * (x_j/L_x)
    return exp( (V_j - V_i) * params['q'] / params['k_B'] / T ) # \Delta energy is -(V_j - V_i) * -params['q']
# the above energies assume a paralell plate capacitor of infinte area. This is hopefully a decent approximation

def boltz_fact_dist(d,Volt, T, L_x):
    '''
    x_i is the current location of the electron and x_j is the location of the electron in the proposed step
    '''
    Delta_V = Volt * (d/L_x)
    return exp( (Delta_V) * params['q'] / params['k_B'] / T ) # \Delta energy is -(V_j - V_i) * -params['q']
# at a temperature of 300 K, params['q'] / params['k_B'] / T = 38.681727071833606
# the above energies assume a paralell plate capacitor of infinte area. This is hopefully a decent approximation



def calculate_current(x, y, L_x, L_y, Volt, lambda_, Rt, T, num_e): # Ben's Monte-Carlo method
    # What this function does:
    # 0) If voltage is 0, return 0 since that will be the end result anyway
    # 1) Introduce num_e electrons at random y-axis locations.
    # 2) for each particle choose to move forward with boltzman probability or backwards so sum is unity
    # 3) Find 3 closest particles in the specified direction
    # 4) Jump to that particle and record the resistence of the jump
    # 5) Interate until we reach the end electrode
    # 6) Sum up the resistances of the jumps
    # 7) Average the resistences together
    # 8) If not too slow - repeat and average again
    # 9) Multiply the final reciprial resistence by the voltage - gives the current; return thus value

    if Volt == 0. or Volt == 0: # step 0
        return 0.

    #Step 1 - introduce electrons on the left electrode
    e_position = hstack([zeros(num_e),L_y * rand(num_e)],dtype=float32) # use float 32 for performance; most of the error comes from integration anyway
    # row 0 are x-values; row 1 are y-values

    x_now = copy(x)
    y_now = copy(y)
    resist = []
    for e in range(num_e): #iterate over all simulated electrons
        x_e, y_e = (e_position[0,e],e_position[1,e])
        resist_path = [] # resitances along the path of a single electron
        while x_e < L_x: # while we have not reached the right electrode
            # step 2: choose direction; use shortest distance to next particle in line to proxy jump distance
            dx = abs(x_now - x_e)
            dx_min = min(dx[abs(dx) < 5e-15])
            Boltzmann_F = boltz_fact_dist(2*dx_min,Volt,T,L_x) # this gives a ratio of occupancies at apprimate closet state to the left vs the appproximate closest state to the right

            # thus we can say that the probability of going right is approximately Boltzmann_F/Boltzmann_B times the probability of going left
            Probability_left = 1./(1. + Boltzmann_F) # probability of moving left
            if rand() > Probability_left:
                Forward = True # choose to move forward with probability according to the electric field energy bias
            else:
                Forward = False

            if Forward: # if moving forward
                x_possible = x[x > x_e]
            else:
                x_possible = x[x < x_e]
                y_possible = y_now[x > x_e]
            # find nearest particle in front of x_i
            d = distances(x_e,x_possible,y_e,y_possible) # distance vector comparing each other point x_j to x_i where x_j > x_i
            # at this point d is an empty array when x_possible is empty
            d = app(d,L_x-x_e) # If the final electrode is closer then jump there instead of a particle.
            j = argmin(d)
            resist_path.append(Rt * exp( d[j] / lambda_))
            if j < len(d)-1:
                e_position[0,e], e_position[1,e] = ( x_possible[j], y_possible[j]) # update the electron positions
            else: # d[j] = L_x - x_e
                e_position[0,e], e_position[1,e] = ( L_x, 0) # y_position doesn't matter at the end electrode
        resist.append( sum(resist_path) ) # in the future I would like to make it so that this chooses between the three loswest values with some weight; for now the nearest neighbor model works
       
    R = sum(resist)/num_e # average the resitances
    return Volt / R


'''def calculate_current(x, y, L, Volt, lambda_, Rt, T, num_e):
    """
    Translated from MATLAB calculateCurrent.m
    Simulates electron hopping and returns current I.

    x is a vector of particle positions (abscissa)
    y is a vector of particle positions (ordinate)
    """
    threshold_dist = 50
    x_ind_sort = argsort(x) # sort the x_values from least to most. Return the inicies
    x_sort = x[x_ind_sort]
    y_sort = y[x_ind_sort]

    # Skip simulation if device too long
    min_distance_to_end = min(L - x) # minimum distance of a particle to the end electrode
    if min_distance_to_end > threshold_dist:
        return 0
    
    # the goal is to estimate the sum R_M \sum_{i=0}^{N-1} R_t * exp( (x_{i+1} - x_i) / lambda ) which is the resistance
    # along one possible path. There are infintely many possible paths. If we consider each possible path a resistor in paralell then
    # the total system resistance is given by $ 1/R_tot = 1/R_{M,1} + 1/R_{M,1} + 1/R_{M,2} + 1/R_{M,3} + ... $.
    # Thus the goal should be to compute this sum and then take the reciprical. This is simular to the problem of computing a partition
    # function for a system with infinite states; we can do it with monte-carlo methods. We do need to decide which paths contribute the most
    # to the final resistance. If we assume that there are $N$ total paths (the paths that we end up sampling) the the total resistence is 
    # this is the resistance along one possible path for the electron.
    # To simplify the number of paths to choose from we use the nearest neighbor model - i.e. the only path the matters is the path between the two closest particles.
    # Other paths will have a much lower change of beign taken. Thus we start with each particle; find the next closest with a higher x-value and so on.
    # Eventually we get to the farthest right particle.
    # This affords us a decent estimate of the resistance of the material.
    # Then the current may be computed with $V = IR \implies I = V/R$ (then inversion of the sum is not needed since we only need 1/R anyway).
    # One last note is that the tunneling probability should be additionally modified by the Boltzmann factor bias that
    # forces migration accordig to the voltage bias. This should allows us to see pulses in the current that mimic pulses in the voltage.

    x_max = x_sort[-1]
    inv_Resists = []
    for i in range(num_e):
        # start an electron at the y-level of each particles' x-value if that particle is one of the num_e closest particles to the left eleectrode
        x_i = x_sort[i]
        y_i = y_sort[i]
        Resist_i = []
        Resist_i.append( Rt * exp( x_i / lambda_) * boltz_fact(0,x_i,Volt,T,L)  )
        while x_i <= x_max:
            x_possible = x[x > x_i]
            # find nearest particle in front of x_i
            d = distances(x_i,x_possible,y_i,y[x > x_i]) # distance vector comparing each other point x_j to x_i where x_j > x_i
            # at this point d is an empty array when x_possible is empty
            d = app(d,L-x_i) # If the final electrode is closer then jump there instead of a particle.
            j = argmin(d)
            
            if j == len(d)-1: # if the electrode is the closest jump; always the case when x = x_max
                Resist_i.append( Rt * exp( d[j] / lambda_) * boltz_fact(x_i,L,Volt,T,L)  )
                x_i = x_max+1 # terminate the while loop
            else:
                x_j = x_possible[j]
                # compute resistance of the tunnel
                Resist_i.append( Rt * exp( d[j] / lambda_) * boltz_fact(x_i,x_j,Volt,T,L)  )
                x_i = copy(x_j)
                y_i = x[j]

        inv_Resists.append( sum(1./array(Resist_i)) )

    inv_R = sum(inv_Resists)

    return inv_R * Volt # return an estimate of the classical current
'''
# below is the old model
'''    LEFT_ELECTRODE = n + 1
    RIGHT_ELECTRODE = n + 2

    # Start all electrons at left electrode
    curr_indexes = full(num_e, LEFT_ELECTRODE) # array of length of the number of simulated electrons filled with n+1
    currx = zeros(num_e)
    curry = full(num_e, nan) # array of length of the number of simulated electrons filled with nan

    for _ in range(steps):
        dx = tile(x, (num_e, 1)) - currx[:, newaxis] # generates a len(x) by num_e matrix. Rows are copies of x minus some value.
        dy = tile(y, (num_e, 1)) - curry[:, newaxis] 
        dy[isnan(dy)] = 0 # on first iteration it sets the dy matrix to 0 - not sure why we don't just set it to zero from the start
        d = sqrt(dx ** 2 + dy ** 2)

        positive = dx >= 0 # Bollean array, returns True at index i if x[i] >=0 and False otherwise
        resist = Rt * exp(d / lambda_) # resistance 
        dist_weights = positive * (1 / resist)

        resist_end = Rt * exp( (L - currx) / lambda_ )
        to_end = 1. / resist_end
        to_end[curr_indexes == RIGHT_ELECTRODE] = 1000  # stay at end
        to_begin = zeros(num_e)

        tot_weights = hstack([dist_weights, to_begin[:, None], to_end[:, None]]) # makes a 3 x 

        curr_indexes, currx, curry = get_next_index(tot_weights, x, y, currx, curry, curr_indexes, L)

    I = sum(currx == L) # after the interations, the current is computed to be the 
    return I'''